cpmpy/expressions/core.py

```diff
@@ -137,12 +159,44 @@ def __repr__(self):
     def __hash__(self):
         return hash(self.__repr__())
 
+    def has_subexpr(self):
+        """ Does it contains nested Expressions (anything other than a _NumVarImpl or a constant)?
+            Is of importance when deciding whether certain transformations are needed
+            along particular paths of the expression tree.
+            Results are cached for future calls and reset when the expression changes
+            (in-place argument update).
+        """
+
+        def recursive_has_subexpr(lst) -> bool:
+            """ Recursive implementation to handle nested lists of expressions.
+            """
+            for el in lst:
+                if isinstance(el, Expression):
+                    if not el.is_leaf():  # NDVarArrays are Expr and any_list, so they are covered too (allows early-exit for decision var arrays)
+                        return True
+                elif is_any_list(el) and recursive_has_subexpr(el): # recursively call on list-like
+                    return True
+            return False
+
+        # If not an Expression (e.g. list-like) or _has_subexpr has not been computed before / has been reset
+        if not hasattr(self, '_has_subexpr'):
+            # args can have lists of lists... -> need for recursive implementation
+            self._has_subexpr = recursive_has_subexpr(self.args)
```

Comment:
hmm, I'm hesitant. I think the 'hot path' should not call a recursive function, just do the loop here.

Then, if an arg is a 'any_list', actually there is one optimisation: if its an NDVarArray and type != object, you don't even need to walk it.
Otherwise... I guess a recursive function, or you do it with a stack, just keep popping till its emtpy (and add to it if you encounter a list)

it might not actually have an impact on efficiency, so still hesitation from my side ; )